### ISSUE TITLE
Block in galacticraft dimensions at 287

### ISSUE THREAD

**Issue by Slind14** [micdoodle8/Galacticraft#2019]
Hey,

this has been a long standing issue that got worse with the last few updates. The highest block in the galacticraft dimensions seems to be different in a way that bukkit doesn't know how to deal with it.

```
Erroring Block:   DIM-41   -7232,287,6680   
java.lang.ArrayIndexOutOfBoundsException: 17
[10:29:57] [Craft Scheduler Thread - 193/WARN]: java.lang.ArrayIndexOutOfBoundsException: 17
[10:29:57] [Craft Scheduler Thread - 193/WARN]:     at org.bukkit.craftbukkit.v1_7_R4.CraftChunkSnapshot.getBlockTypeId(CraftChunkSnapshot.java:55)
```

Is there some special block at this height? When I look into the world at this height, there doesn't appear to be anything.

This includes addons like 4Space.

**Comment by Slind14** [comment]
Edit: sorry, I believe I was wrong. It happened before with galacticraft only packs, but in this case it doesn't. Might have been changed/fixed since. 
If you know anything about 4Space and this type of issue I would appreciate any input.

**Comment by radfast** [comment]
I can't help with 4Space.  You would need to contact the author with any issues.  It is not supported through this repository - I have never seen its source code and don't know too much about it.  

Some of the other GC add-on authors participate actively in this repository and their contributions are greatly appreciated.  But there's a kind of arm's length relationship with the 4Space author, some kind of drama a couple of years back where he infringed GC's copyright or didn't do what micdoodle8 asked him, I'm not too sure of details as I wasn't around in those days.


----

My thoughts: I don't think a block in the world can have a y-value of 287 unless you have installed a mod allowing extended chunks.  The vanilla save format, used by Forge, is limited to blocks in the 0-255 range.  But I can easily imagine code calling - for example - ```world.getBlock(x, y, z)``` with ```y``` outside the 0 to 255 range.  I'm sure Galacticraft's own code sometimes does that when testing the blocks above or below a tile.

A y-value of 287 suggests something may be looking to see what the block is 32 blocks above itself.  Hard to imagine what that could be - maybe some kind of worldgen which generates 32-block deep shafts or 32-block tall towers?  It could also be, for example, a force field or other distance effect with a 32-block radius; a nuke from ICBM mod or similar large explosion; or somebody using WorldEdit to move something 32 blocks up or down.

Just because the error is in dimension -41, does not at all mean that it is the mod which provides dimension -41 which is erroring.  (That would be like the players who come on here saying, "GC crashes when I login to my Moon Base" and then we find that their moon base contains blocks, tileentities and entities from 30 other mods.)

Actually I would guess that if a CraftBukkit class is erroring, then it must be a Bukkit plugin which is triggering the error.

**Comment by Slind14** [comment]
Alright thanks. I noticed my error, we had similar issue before which only affected galcticraft worlds (no galacticraft addons installed), so I though it was back, sorry to trouble you about something unrelated. 

Might I ask if you know a good 4Space alternative that you could recommend for https://github.com/MyM-ModpackTeam/GalacticScience

**Comment by radfast** [comment]
Nice modpack, I hadn't looked at it before.

I personally don't play the AddOns as my focus is on coding GC itself (I don't really *play* Minecraft much at all these days).  You should ask that question on the forums, you will get a lot of comments there.

AFAIAA the main ones are:

* MorePlanets from @SteveKunG which is open source and offers both known planets in the solar system and a few more colourful / fantasy type planets round nearby star systems.  You might be able to disable the fantasy planets in configs, you could talk to him about that?

* A closed source mod from @BlesseNtumble which you can see in a screenshot at https://github.com/micdoodle8/Galacticraft/issues/1557 - you don't need to know much Russian to see that that has Mercury, Venus, and Europa added.  All the screenshots I've seen are in the Russian language, I don't know if his mod has an English translation - but it might need only 30 minutes to translate it.  Again you should ask him about that directly.

I've seen mention of a few other AddOns in current / recent discussions on the forum - there is an AddOns area - but I have no idea how complete or close to release they are.  There's a general pattern of people starting work on an AddOn with exciting plans for it and great intentions, but then finding that coding and debugging it is way harder than they expected, so it stops...

My general impression is that AddOns typically re-use the worldgen code from Galacticraft's Moon and Mars, of course with different coloured blocks and different sky, and sometimes changing the hill parameters etc.  This results in similar-looking planets, different colours but not very interesting to play on.  It kind of cheapens all the hard work which has gone into making the Moon and Mars look so different from the Overworld.  I think some of the AddOns are gradually working on making their worlds more distinct, but you will have to try them out and judge for yourself.

TL;DR: writing good worldgen code, which makes an interesting world to play in, is hard.

----

Later in 2016 if all goes well I plan to: (1) revisit the Mars worldgen code in Galacticraft and add more realistic features like canyons, boulder plains, and other Mars-type stuff + (2) code a completely new worldgen for another celestial body which will look different from the Moon and Mars.  Based on that experience, the goal will then be (3) to try to generalise and parameterise more worldgen code and add it to the API, to help AddOn authors more easily create a wider variety of planets.  No guarantees that any of this will happen, due to real life issues or decisions to take the mod in a different direction.

Ultimately it would be really good if GC could algorithmically generate [a whole universe of random planets](http://www.no-mans-sky.com/about/).  But that is a tough call, the type of thing which would take a large team of programmers and designers a year to produce.

To save work, another approach could be to find an existing open-source mod which either changes the vanilla Minecraft worldgen a lot, or creates a new dimension with its own detailed worldgen, and turn that into a planet instead.  I'm thinking specifically of the *Promised Land* dimension in Biomes O' Plenty which has been removed from that mod in Minecraft 1.7.10.  (Other detailed worldgen I like are Tropicraft and Twilight Forest, also of course the new Overworld biomes in Biomes O' Plenty.)  I am not currently working on this, and have not heard of anyone who is, but it's certainly an idea...

None of this discussion about the future helps you now, but just so you know what is planned :)

I like to write this stuff out anyhow, in case it inspires someone else to start contributing ...

**Comment by radfast** [comment]
Later in 2016 if all goes well I plan to: (1) revisit the Mars worldgen code in Galacticraft and add more realistic features like canyons, boulder plains, and other Mars-type stuff + (2) code a completely new worldgen for another celestial body which will look different from the Moon and Mars.  Based on that experience, the goal will then be (3) to try to generalise and parameterise more worldgen code and add it to the API, to help AddOn authors more easily create a wider variety of planets.

No guarantees that any of this will happen, due to real life issues or decisions to take the mod in a different direction.

Ultimately it would be really good if GC could algorithmically generate [a whole universe of random planets](http://www.no-mans-sky.com/about/).  But that is a tough call, the type of thing which would take a large team of programmers and designers a year to produce.  So it will probably never happen.

To save work, another approach could be to find an existing open-source mod which either changes the vanilla Minecraft worldgen a lot, or creates a new dimension with its own detailed worldgen, and turn that into a planet instead.  I'm thinking specifically of the *Promised Land* dimension in Biomes O' Plenty which has been removed from that mod in Minecraft 1.7.10.  (Other detailed worldgen I like are Tropicraft and Twilight Forest, also of course the new Overworld biomes in Biomes O' Plenty.)  I am not currently working on this, and have not heard of anyone who is, but it's certainly an idea...

None of this discussion about the future helps you now, but just so you know what is planned :)

If any of this inspires someone new to start contributing to GC development, feel free :)

**Comment by Slind14** [comment]
Wow, thanks for all the details. Atm we are using the 4Space planets as part of the tiering system. I'm not sure how many players expended to a 4Space planet, so removing it might not be a good idea until a 1.8 version.

@radfast It would be an honor if you gave galactic science a try. Galacticraft is together with minechem the core mod. If you played modded during tekkit times, you might like that it doesn't have any applied energistics and the required amount of resources forces you to build factories :)
If you do, stick to the quest book, else you will waste days on tier 1 :D

**Comment by BlesseNtumble** [comment]
Galaxy Space full on English lang.

![1](http://firepic.org/images/2016-01/15/2h5zx4ds4j7h.png)
![2](http://firepic.org/images/2016-01/15/jdj0otm372my.png)
![3](http://firepic.org/images/2016-01/15/vhk4j8vq2kdu.png)

**Comment by radfast** [comment]
Nice.  And nice 'alpha'. :)

**Comment by radfast** [comment]
I think GalacticSpace modpack should use GalaxySpace mod :)
